src/internal/operators/shareReplay.ts

```diff
@@ -168,19 +167,24 @@ function shareReplayOperator<T>({
           subject!.error(err);
         },
         complete() {
-          isComplete = true;
           subscription = undefined;
           subject!.complete();
         },
       });
+      // The following condition is needed because source can complete synchronously
+      // upon subscription. When that happens `subscription` is first set to `undefined`
+      // and right after is set to the "closed subscription" returned by `subscribe`
+      if (subscription.closed && !hasError) {
```

Comment:
It shouldn't be necessary to check `hasError`, here. Not nulling-out the `subscription` in the error handler was an oversight. AFAICT, there's no reason the `subscription` should not be set to `undefined` when an error occurs.

```ts
        error(err) {
          hasError = true;
          subscription = undefined;
          subject!.error(err);
        },
```

Reply:
You are absolutely right. In fact I also wanted to release the subject in there too, that way we would avoid a possible memory-leak and also the `hasError` flag would no longer be needed... But I thought: "touch as little as possible, because the last time you already broke it".

On the other hand, thanks to the fact that I "broke" it, we became aware of this annoyance, so this time I've decided to be a bit braver and I've done the refactor that I think that this operator needed.

Reply:
@cartant have you had a chance to re-review the latest changes? Also, it goes without saying that I don't mind closing this PR if you rather not touch this file unless there is an actual need/bug.

Reply:
I'll get to it eventually.